### PR TITLE
Enable Coretime on Polkadot Relay

### PR DESCRIPTION
**Pull request by tdimitrov**
Redo of https://github.com/polkadot-fellows/runtimes/pull/354 on top of https://github.com/polkadot-fellows/runtimes/pull/381

TODOs:
- [ ] Check migrations with chopsticks
- [ ] Regenerate weights

### CODE REVIEW

File: relay/polkadot/src/lib.rs [hunk 1]
```diff
@@ -1780,12 +1832,94 @@ pub type Migrations = (migrations::Unreleased, migrations::Permanent);
...
+	/// This is required for Coretime to ensure the relay chain processes parachains that are
+	/// assigned to multiple cores.
+	pub struct EnableElasticScalingNodeFeature;
+	impl OnRuntimeUpgrade for EnableElasticScalingNodeFeature {
+		fn on_runtime_upgrade() -> Weight {
+			use runtime_parachains::configuration::WeightInfo as _;
+			let _ = Configuration::set_node_feature(frame_system::RawOrigin::Root.into(), 1, true);
+			weights::runtime_parachains_configuration::WeightInfo::<Runtime>::set_node_feature()
+		}
+	}
```
Comment (alindima) [inline relay/polkadot/src/lib.rs:1883]: no longer needed, thanks to https://collectives.subsquare.io/fellowship/referenda/173

File: relay/polkadot/src/lib.rs [hunk 2]
```diff
@@ -1780,12 +1832,81 @@ pub type Migrations = (migrations::Unreleased, migrations::Permanent);
...
+		fn on_runtime_upgrade() -> Weight {
+			if let Err(err) = Auctions::cancel_auction(frame_system::RawOrigin::Root.into()) {
+				log::debug!(target: "runtime", "Cancelling auctions failed: {:?}", err);
+			}
+			// Cancel scheduled auction as well:
+			if let Err(err) = Scheduler::cancel_named(
+				// TODO: Fix this key for Polkadot!
+				pallet_custom_origins::Origin::AuctionAdmin.into(),
+				[
+					0x5c, 0x68, 0xbf, 0x0c, 0x2d, 0x11, 0x04, 0x91, 0x6b, 0xa5, 0xa4, 0xde, 0xe6,
```
Comment (tdimitrov) [inline relay/polkadot/src/lib.rs:1894]: I assume this key is for scheduler -> lookup, right?

image

How can I understand which is the correct key?

Comment (eskimor) [reply relay/polkadot/src/lib.rs:1894]: I have the right key already.

Comment (eskimor) [reply relay/polkadot/src/lib.rs:1894]: I found it by searching for `Origins: AuctionAdmin`. It is:

                [ 0x87, 0xa8, 0x71, 0xb4, 0xd6, 0x21, 0xf0, 0xb9, 0x73, 0x47, 0x5a, 0xaf, 0xcc,
                0x32, 0x61, 0x0b, 0xd7, 0x68, 0x8f, 0x15, 0x02, 0x33, 0x8a, 0xcd, 0x00, 0xee, 0x48,
                0x8a, 0xc3, 0x62, 0x0f, 0x4c, ],

Comment (eskimor) [reply relay/polkadot/src/lib.rs:1894]: I'll provide a PR with this and other changes against your PR.

File: CHANGELOG.md [hunk 3]
```diff
@@ -15,6 +15,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Core-Fellowship: new `promote_fast` call ([runtimes#381](https://github.com/polkadot-fellows/runtimes/pull/381), [SDK v1.14 #4877](https://github.com/paritytech/polkadot-sdk/pull/4877)).
 - Pallet ranked collective: max member count per rank ([runtimes#381](https://github.com/polkadot-fellows/runtimes/pull/381), [SDK v1.14 #4807](https://github.com/paritytech/polkadot-sdk/pull/4807)).
 - Pallet assets: optional auto-increment for the asset ID ([runtimes#381](https://github.com/polkadot-fellows/runtimes/pull/381), [SDK v1.14 #4757](https://github.com/paritytech/polkadot-sdk/pull/4757)).
+- Enable Agile Coretime on Polkadot ([polkadot-fellows/runtimes#322](https://github.com/polkadot-fellows/runtimes/pull/322))
+- Enable Elastic Scaling node side feature for Polkadot ([polkadot-fellows/runtimes#322](https://github.com/polkadot-fellows/runtimes/pull/322))
```
Comment (alindima) [inline CHANGELOG.md:19]: also remove from changelog

File: relay/polkadot/src/lib.rs (hunk 2 above)
Comment (tdimitrov) [reply relay/polkadot/src/lib.rs:1894]: Feel free to commit in this PR directly (I've invited you as a collaborator).

File: relay/kusama/src/lib.rs [hunk 4]
```diff
@@ -1677,7 +1677,7 @@ construct_runtime! {
 		ParaSessionInfo: parachains_session_info = 61,
 		ParasDisputes: parachains_disputes = 62,
 		ParasSlashing: parachains_slashing = 63,
-		OnDemandAssignmentProvider: parachains_assigner_on_demand = 64,
+		OnDemand: parachains_assigner_on_demand = 64,
```
Comment (s0me0ne-unkn0wn) [inline relay/kusama/src/lib.rs:1680]: This renaming requires a storage migration using https://github.com/paritytech/polkadot-sdk/blob/604f56f03db847a90aa4fdb13be6b80482a4dcd6/substrate/frame/support/src/storage/migration.rs#L323
As storage keys will change

File: CHANGELOG.md (hunk 3 above)
Comment (tdimitrov) [reply CHANGELOG.md]: Thanks Alin. Fixed.

File: relay/kusama/src/lib.rs (hunk 4 above)
Comment (eskimor) [reply relay/kusama/src/lib.rs:1680]: @tdimitrov Can you look into that please?

**Comment by eskimor** [comment]
Collecting things to test/check after that PR is ready:

- [ ] Migrations ran successfully with no warnings or errors
- [ ] Cancel auction actually worked (schedule got updated, key no longer present)
- [ ] All leases correctly present on Coretime chain, including those that have not yet started. (they exist, they end at the right time?)
- [ ] In general all assignments exist on the coretime chain: System chains + Leases
- [ ] Relay chain state matches coretime state
- [ ] We have one on-demand core

**Review by eskimor — APPROVED** [review]

File: relay/polkadot/src/lib.rs [hunk 5]
```diff
@@ -1780,22 +1841,102 @@ pub type Migrations = (migrations::Unreleased, migrations::Permanent);
...
+			move_storage_from_pallet(b"Pallet", b"OnDemandAssignmentProvider", b"OnDemand");
+			move_storage_from_pallet(b"ParaIdAffinity", b"OnDemandAssignmentProvider", b"OnDemand");
+			move_storage_from_pallet(b"QueueStatus", b"OnDemandAssignmentProvider", b"OnDemand");
+			move_storage_from_pallet(b"FreeEntries", b"OnDemandAssignmentProvider", b"OnDemand");
+			move_storage_from_pallet(
+				b"AffinityEntries",
+				b"OnDemandAssignmentProvider",
+				b"OnDemand",
+			);
+			move_storage_from_pallet(b"Revenue", b"OnDemandAssignmentProvider", b"OnDemand");
```
Comment (tdimitrov) [inline relay/polkadot/src/lib.rs:1917]: @s0me0ne-unkn0wn does this look right to you?

As far as I understand I need to call `move_storage_from_pallet` for each storage item in the pallet?
Are the pallet names correct?

Ideally I should use `get_storage_prefix()` here but the storage items are not accessible here so I hardcoded them :1st_place_medal:

Comment (s0me0ne-unkn0wn) [reply relay/polkadot/src/lib.rs:1859]: > @s0me0ne-unkn0wn does this look right to you?

Well, it looks totally fine to me, but it should be taken into account that I didn't ever do it myself as well tbh :laughing:

> As far as I understand I need to call `move_storage_from_pallet` for each storage item in the pallet?
> Are the pallet names correct?

My understanding is the same, and the naming seems to be fine. I still hope to get some reviews from guys who have closer experience :slightly_smiling_face:

File: relay/polkadot/src/lib.rs [hunk 6]
```diff
@@ -1726,13 +1786,93 @@ pub type Migrations = (migrations::Unreleased, migrations::Permanent);
...
+			weights::polkadot_runtime_common_auctions::WeightInfo::<Runtime>::cancel_auction()
+				.saturating_add(weights::pallet_scheduler::WeightInfo::<Runtime>::cancel_named(
+					<Runtime as pallet_scheduler::Config>::MaxScheduledPerBlock::get(),
+				))
+		}
+	}
+
+	// Migrate storage for pallet rename `OnDemandAssignmentProvider` -> `OnDemand`
+	pub struct OnDemandRename;
+	impl OnRuntimeUpgrade for OnDemandRename {
```
Comment (eskimor) [inline relay/polkadot/src/lib.rs:1847]: Two things here:

1. For Polkadot no migration should be needed as onDemand was newly introduced in this PR.
2. Do we need any protection here from preventing this migration to be run multiple times?

**Review by eskimor — CHANGES_REQUESTED** [review]

**Comment by tdimitrov** [comment]
I've extracted the pallet rename on Kusama in a separate PR: https://github.com/polkadot-fellows/runtimes/pull/417

File: relay/polkadot/src/lib.rs [hunk 7]
```diff
@@ -1726,13 +1786,73 @@ pub type Migrations = (migrations::Unreleased, migrations::Permanent);
...
+
 	use super::*;
 
+	pub struct GetLegacyLeaseImpl;
+	impl coretime::migration::GetLegacyLease<BlockNumber> for GetLegacyLeaseImpl {
+		fn get_parachain_lease_in_blocks(para: ParaId) -> Option<BlockNumber> {
+			let now = frame_system::Pallet::<Runtime>::block_number();
+			let lease = slots::Leases::<Runtime>::get(para);
+			if lease.is_empty() {
+				return None
```
Comment (seadanda) [inline relay/polkadot/src/lib.rs:1799]:
```suggestion
				return None;
```

**Review by seadanda — APPROVED** [review]
Looks good

File: relay/polkadot/src/lib.rs [hunk 8]
```diff
@@ -1374,16 +1378,63 @@ impl parachains_paras_inherent::Config for Runtime {
...
-	type AssignmentProvider = ParaAssignmentProvider;
+	// If you change this, make sure the `Assignment` type of the new provider is binary compatible,
+	// otherwise provide a migration.
+	type AssignmentProvider = CoretimeAssignmentProvider;
+}
+
+parameter_types! {
+	pub const BrokerId: u32 = system_parachain::BROKER_ID;
+	pub const BrokerPalletId: PalletId = PalletId(*b"py/broke");
+	pub MaxXcmTransactWeight: Weight = Weight::from_parts(constants::WEIGHT_REF_TIME_PER_MILLIS, 20 * constants::WEIGHT_PROOF_SIZE_PER_KB);
```
[the remark this answers is not in the document]
Comment (seadanda) [reply relay/polkadot/src/lib.rs:1389]: https://github.com/polkadot-fellows/runtimes/blob/main/relay/kusama/src/lib.rs#L1288

[the remark this answers is not in the document]
Comment (tdimitrov) [reply relay/polkadot/src/lib.rs]: My bad, I probably took the values from master instead of from the release branch. Fixed.

**Review by eskimor — APPROVED** [review]

File: relay/polkadot/src/weights/runtime_parachains_coretime.rs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+//! DATE: 2024-03-10, STEPS: `50`, REPEAT: `20`, LOW RANGE: `[]`, HIGH RANGE: `[]`
+//! WORST CASE MAP SIZE: `1000000`
+//! HOSTNAME: `ggwpez-ref-hw`, CPU: `Intel(R) Xeon(R) CPU @ 2.60GHz`
+//! WASM-EXECUTION: `Compiled`, CHAIN: `Some("./kusama-chain-spec.json")`, DB CACHE: 1024
+
+// Executed Command:
+// ./target/production/polkadot
+// benchmark
+// pallet
+// --chain=./kusama-chain-spec.json
```
Comment (sandreim) [inline relay/polkadot/src/weights/runtime_parachains_coretime.rs:28]: Same question as above

File: relay/polkadot/src/weights/runtime_parachains_assigner_on_demand.rs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+//! DATE: 2024-03-10, STEPS: `50`, REPEAT: `20`, LOW RANGE: `[]`, HIGH RANGE: `[]`
+//! WORST CASE MAP SIZE: `1000000`
+//! HOSTNAME: `ggwpez-ref-hw`, CPU: `Intel(R) Xeon(R) CPU @ 2.60GHz`
+//! WASM-EXECUTION: `Compiled`, CHAIN: `Some("./kusama-chain-spec.json")`, DB CACHE: 1024
+
+// Executed Command:
+// ./target/production/polkadot
+// benchmark
+// pallet
+// --chain=./kusama-chain-spec.json
```
Comment (sandreim) [inline relay/polkadot/src/weights/runtime_parachains_assigner_on_demand.rs:28]: Shouldn't this be generated with --chain=./polkadot-chain-spec.json ?

**Review by acatangiu — APPROVED** [review]

File: relay/polkadot/src/lib.rs [hunk 11]
```diff
@@ -1506,6 +1563,13 @@ impl pallet_asset_rate::Config for Runtime {
 	#[cfg(feature = "runtime-benchmarks")]
 	type BenchmarkHelper = polkadot_runtime_common::impls::benchmarks::AssetRateArguments;
 }
+// Notify `coretime` Pallet when a lease swap occurs
```
Comment (joepetrowski) [inline relay/polkadot/src/lib.rs:1566]:
```suggestion
/// Notify the `coretime` pallet when a lease swap occurs.
```

**Review by joepetrowski — APPROVED** [review]

File: relay/polkadot/src/weights/runtime_parachains_assigner_on_demand.rs (hunk 10 above)
Comment (joepetrowski) [reply relay/polkadot/src/weights/runtime_parachains_assigner_on_demand.rs:28]: There should be another pre-release PR that regenerates all the weights. Usually @bkontur does it (although would be good to improve/automate that process)

**Pull request merged**